### PR TITLE
add the paddle.distributed.split api

### PR DESCRIPTION
**Pull request by sandyhouse**
### PR types

New features
### PR changes

APIs
### Describe

Add the paddle.distributed.split api, which implements the parallel embedding and parallel linear.
It splits the weight of the specified operation into multiple devices and do the computation in parallel.

Now the following three cases are supported.

- Case 1: Parallel Embedding
   The weight of the embedding operation is a NxM matrix with N rows and M columns. With parallel embedding, the weight is split into num_partitions partitions, each of which is a matrix with N/num_partitions rows and M column.
   Usage：
```python
import numpy
import paddle
from paddle.distributed import init_parallel_env
init_parallel_env()
np_data = numpy.random.randint(0, 8, (10,4))
data = paddle.to_tensor(np_data)
emb_out = padle.distributed.split(
    data,
    (8, 128)
    operation="embedding",
    num_partitions=2)
```

- Case 2: Row Parallel Linear
   The weight of the linear operation is a NxM matrix with N rows and M columns. With row parallel linear, the weight is split into num_partitions partitions, each of which is a matrix with N/num_partitions rows and M column.
   Usage：
```python
import numpy
import paddle
from paddle.distributed import init_parallel_env
init_parallel_env()
np_data = numpy.random.rand(10, 1000)
data = paddle.to_tensor(np_data)
data = paddle.distributed.broadcast(data, src=0)
rank = paddle.distributed.ParallelEnv().local_rank
data = paddle.split(data, 2, axis=1)[rank]
linear_out = padle.distributed.split(
    data,
    (1000, 16)
    axis=0,
    operation="linear",
    num_partitions=2)
```

- Case 3: Column Parallel Linear
   The weight of the linear operation is a NxM matrix with N rows and M columns. With column parallel linear, the weight is split into num_paratitions partitions, each of which is a matrix with N rows and M/num_partitions column.
   Usage：
```python
import numpy
import paddle
from paddle.distributed import init_parallel_env
init_parallel_env()
np_data = numpy.random.rand(10, 1000)
data = paddle.to_tensor(np_data)
data = paddle.distributed.broadcast(data, src=0)
linear_out = padle.distributed.split(
    data,
    (1000, 16)
    axis=1,
    operation="linear",
    num_partitions=2)
```

TODO (sandyhouse) : 
- Add performance and benefits of the use of this api.
- Do full tests on dynamic graph

image

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -485,3 +487,190 @@ def barrier(group=0):
...
+    else:
+        assert fleet._role_maker, ("To use paddle.distributed.split, "
+                                   "you must call fleet.init() firstly.")
+        rank = fleet.worker_index()
+        nranks = fleet.worker_num()
+
+    # rank within a model parallel group
+    inner_rank = rank % num_partitions
+
+    if operation == "embedding":
```
Comment (gongweibao) [inline python/paddle/distributed/collective.py:667]: 这个两个if, else太大， 封装为函数吧！

File: python/paddle/distributed/collective.py [hunk 2]
```diff
@@ -485,3 +487,190 @@ def barrier(group=0):
...
+        With row parallel linear, the weight is split into num_partitions partitions, each
+        of which is a matrix with N/num_partitions rows and M column.
+
+    Case 3: Column Parallel Linear
+        The weight of the linear operation is a NxM matrix with N rows and M columns.
+        With column parallel linear, the weight is split into num_paratitions partitions, each
+        of which is a matrix with N rows and M/num_partitions column.
+
+    Args:
+        x (Tensor): Input tensor. It's data type should be float16, float32, float64, int32 or int64.
```
Comment (gongweibao) [inline python/paddle/distributed/collective.py:608]: 这个可能需要更好的描述。

File: python/paddle/distributed/collective.py [hunk 3]
```diff
@@ -485,3 +487,190 @@ def barrier(group=0):
...
+        rank = fleet.worker_index()
+        nranks = fleet.worker_num()
+
+    # rank within a model parallel group
+    inner_rank = rank % num_partitions
+
+    if operation == "embedding":
+        assert axis == 0, ("We only support to split the weight of embedding "
+                           "along the first axis now.")
+        assert size[0] % num_partitions == 0, (
```
Comment (gongweibao) [inline python/paddle/distributed/collective.py]: embedding的切分应该不用均分？

**Review by gongweibao — COMMENTED** [review]
1.这个或者下个PR加一下动态图的单测？看API的调用是动静都支持的。
2.描述部分缺性能和单Feature最大模型的测试。
3.Save/Load模型的方式？

File: python/paddle/distributed/collective.py [hunk 4]
```diff
@@ -485,3 +487,190 @@ def barrier(group=0):
...
+
+            if not name:
+                name = "fc_by_col_rank_%d" % inner_rank
+            else:
+                name = name + "_by_col_rank_%d" % inner_rank
+        else:
+            raise ValueError("The value of axis must be 0 or 1, but the value "
+                             "given is {}.".format(axis))
+
+        linear = paddle.nn.Linear(
```
Comment (gongweibao) [inline python/paddle/distributed/collective.py]: x是怎么切的？

File: python/paddle/distributed/collective.py (hunk 1 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py:667]: Done.

File: python/paddle/distributed/collective.py (hunk 2 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py:608]: Done. Add more details.

File: python/paddle/distributed/collective.py (hunk 3 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py:671]: 目前只支持均分的配置。如果不均分的话最后需要对结果做复杂的padding操作，已执行all_reduce_sum。

File: python/paddle/distributed/collective.py (hunk 4 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: 目前由用户切分x。若用户未进行切分会给出错误提示信息。

File: python/paddle/distributed/collective.py (hunk 3 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done. 支持不均匀切分并增加相应单测。

**Comment by sandyhouse** [comment]
> 1.这个或者下个PR加一下动态图的单测？看API的调用是动静都支持的。
> 2.描述部分缺性能和单Feature最大模型的测试。
> 3.Save/Load模型的方式？

1、2增加了TODO；
3. 针对Save可以切分后的参数单独保存，例如embedding参数切分为2部分后，Save的时候两个部分的参数都会保存，再Load时分别加载。有一种特殊情形，加载时和保存时切分数量不同。对这种特殊情形可以提供额外的工具做转换。后续可以讨论是否有更好的实现方案。

File: python/paddle/distributed/collective.py [hunk 5]
```diff
@@ -485,3 +487,228 @@ def barrier(group=0):
...
+    return emb_out
+
+
+def split(x,
+          size,
+          operation,
+          axis=0,
+          num_partitions=1,
+          gather_out=True,
+          param_attr=None,
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: param_attr -> weight_attr

File: python/paddle/distributed/collective.py [hunk 6]
```diff
@@ -485,3 +487,228 @@ def barrier(group=0):
...
+    Case 3: Column Parallel Linear
+        The weight of the linear operation is a NxM matrix with N rows and M columns.
+        With column parallel linear, the weight is split into num_paratitions partitions, each
+        of which is a matrix with N rows and M/num_partitions column.
+
+    Args:
+        x (Tensor): Input tensor. It's data type should be float16, float32, float64, int32 or int64.
+        size (list|tuple): A list or tuple with two elements indicating the shape of the weight.
+        operation (str): The name of the operation. The supported operations are 'linear' and 'embedding'.
+        axis (int, Optional): Indicate along which axis to split the weight.
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: ``axis、num_partitions 、gather_out param_attr (weight_attr)、bias_attr、name ``没有说明默认值

File: python/paddle/distributed/collective.py [hunk 7]
```diff
@@ -485,3 +487,228 @@ def barrier(group=0):
...
+        num_partitions (int, Optional): How many parts the weight is partitioned.
+        gather_out (bool, Optional): Whether to gather the output after computation.
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: bool型参数需要给出默认值下的行为

File: python/paddle/distributed/collective.py [hunk 8]
```diff
@@ -485,3 +487,228 @@ def barrier(group=0):
...
+        name (str, Optional): The default value is None. Normally there is no need for user to set this
+            property. For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor.
+
+    Examples:
+        .. code-block:: python
+
+            import numpy
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: 可以不用numpy

File: python/paddle/distributed/collective.py [hunk 9]
```diff
@@ -485,3 +487,228 @@ def barrier(group=0):
...
+            import paddle
+            from paddle.distributed import init_parallel_env
+
+            paddle.set_device('gpu:%d'%paddle.distributed.ParallelEnv().dev_id)
+            init_parallel_env()
+            np_data = numpy.random.randint(0, 8, (10,4))
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: paddle.randint 就可以

File: python/paddle/distributed/collective.py (hunk 5 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

File: python/paddle/distributed/collective.py (hunk 6 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

File: python/paddle/distributed/collective.py (hunk 7 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

File: python/paddle/distributed/collective.py (hunk 8 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

File: python/paddle/distributed/collective.py (hunk 9 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Review by gongweibao — DISMISSED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by chalsliu — APPROVED** [review]

**Pull request merged**